File: cwsfroster/ShellViewController.m
//
//  ShellViewController.m
//  cwsfroster
//
//  Created by Bobby Ren on 8/31/14.
//  Copyright (c) 2014 Bobby Ren. All rights reserved.
//

#import "ShellViewController.h"

@interface ShellViewController ()

@end

@implementation ShellViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.

    // special tkd stuff
    [self updateTabBarIcons];
    [self listenFor:@"organization:name:changed" action:@selector(updateTabBarIcons)];
    [self listenFor:@"goToSettings" action:@selector(goToSettings:)];
    
    if ([[NSUserDefaults standardUserDefaults] boolForKey:@"organization:is:new"]) {
        [self setSelectedIndex:1];
    }
}

-(void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    if ([[NSUserDefaults standardUserDefaults] boolForKey:@"organization:is:new"]) {
        [[NSUserDefaults standardUserDefaults] setBool:NO forKey:@"organization:is:new"];
        [[NSUserDefaults standardUserDefaults] synchronize];
        
        NSString *title = [NSString stringWithFormat: @"Welcome to %@", [[Organization currentOrganization] name]];
        [self simpleAlert:title message:@"Add some members to your new organization."];
    }
}

-(void)updateTabBarIcons {
    NSString *name = [[Organization currentOrganization].name lowercaseString];
    if (name &&
        ([[[Organization currentOrganization].name lowercaseString] rangeOfString:@"taekwondo"].location != NSNotFound || [[[Organization currentOrganization].name lowercaseString] rangeOfString:@"tkd"].location != NSNotFound || [[[Organization currentOrganization].name lowercaseString] rangeOfString:@"tae kwon do"].location != NSNotFound)) {
        [self setIcon:@"icon-tkd-paddle" forTabBar:0];
        [self setIcon:@"icon-tkd-helmet" forTabBar:1];
    }
    else {
        [self setIcon:@"icon-calendar" forTabBar:0];
        [self setIcon:@"icon-users" forTabBar:1];
    }
}

-(void)setIcon:(NSString *)iconName forTabBar:(int)tabBarIndex {
    UIViewController * controller = self.viewControllers[tabBarIndex];
    UITabBarItem *item = controller.tabBarItem;
    item.image = [UIImage imageNamed:iconName];
    item.selectedImage = item.image;
}

-(void)goToSettings:(NSNotification *)n {
    UINavigationController *nav = [[UIStoryboard storyboardWithName:@"Settings" bundle:nil] instantiateViewControllerWithIdentifier:@"SettingsNavigationController"];
    [self presentViewController:nav animated:YES completion:nil];
}
- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: cwsfroster/ShellViewController.swift
//
//  ShellViewController.swift
//  rollcall
//
//  Created by Ren, Bobby on 6/1/18.
//  Copyright © 2018 Bobby Ren. All rights reserved.
//

import UIKit
import RxSwift
import Firebase

class ShellViewController: UITabBarController {
    var disposeBag: DisposeBag? = DisposeBag()
    fileprivate var upgradeShown: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        listenFor("organization:name:changed", action: #selector(updateTabBarIcons), object: nil)
        listenFor("goToSettings", action: #selector(goToSettings), object: nil)
        listenFor(.LogoutSuccess, action: #selector(didLogout), object: nil)

        if UserDefaults.standard.bool(forKey: "organization:is:new") {
            selectedIndex = 1
        }
        
        updateTabBarIcons()
        
        listenForOrganization()
    }
    
    deinit {
        // FIXME: for some reason, presenting ShellViewController on SplashViewController causes ShellViewController to never deallocate. maybe it's because of the mix of objc and swift classes? As a result, disposeBag is never deallocated, and listeners and observers never stop observing. We have to force that to happen on logout
        print("deinit succeess")
    }
    
    @objc func didLogout() {
        // this causes listenForOrganization to be successfully cleared even if ShellViewController is not actually correctly deallocated on logout (corner case)
        disposeBag = nil
        print("here didlogout")
        stopListeningFor("organization:name:changed")
        stopListeningFor("goToSettings")
        stopListeningFor(.LogoutSuccess)
        
        UpgradeService().clearOnLogout()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        if UserDefaults.standard.bool(forKey: "organization:is:new") {
            UserDefaults.standard.set(false, forKey: "organization:is:new")
            UserDefaults.standard.synchronize()
            
            if let name = OrganizationService.shared.current.value?.name {
                let title = "Welcome to \(name)"
                simpleAlert(title, message: "Add some members to your new organization")
            }
        }
        
        UpgradeService().promptForUpgradeIfNeeded(from: self)
    }

    func listenForOrganization() {
        print("Listening for organization")
        guard let disposeBag = disposeBag else { return }
        OrganizationService.shared.current.asObservable().distinctUntilChanged().subscribe(onNext: { [weak self] (org) in
            print("Listening for organization -> title: \(org?.name)")
        }).disposed(by: disposeBag)
    }
    
    @objc func updateTabBarIcons() {
        guard let name = OrganizationService.shared.current.value?.name else { return }
        if name.lowercased().contains("taekwondo") {
            setIcon(iconName: "icon-tkd-paddle", for: 0)
            setIcon(iconName: "icon-tkd-helmet", for: 1)
        } else {
            setIcon(iconName: "icon-calendar", for: 0)
            setIcon(iconName: "icon-users", for: 1)
        }
    }

    func setIcon(iconName: String, for index: Int) {
        guard let controller = viewControllers?[index] else { return }
        let image = UIImage(named: iconName)
        controller.tabBarItem.image = image
        controller.tabBarItem.selectedImage = image
    }
    
    @objc func goToSettings() {
        if let nav = UIStoryboard(name: "Settings", bundle: nil).instantiateInitialViewController() {
            present(nav, animated: true, completion: nil)
        }
    }
}
